---
title: "The hummingbird"
author: Eleni Adam
affiliation: Old Dominion University, Norfolk, VA, USA
package: hummingbird
output:
  BiocStyle::html_document:
    toc: true
    toc_depth: 2
vignette: >
  %\VignetteIndexEntry{hummingbird}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=TRUE}
library(knitr)
library(rmarkdown)
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

hummingbird is a package for identifying differentially methylated regions (DMRs) between case and
control groups using whole genome bisulfite sequencing (WGBS) or reduced representative bisulfite
sequencing (RRBS) experiment data.

The hummingbird package uses a Bayesian hidden Markov model (HMM) for detecting DMRs. It fits a 
Bayesian HMM for each chromosome. The final output of hummingbird is the DMRs with start and end 
position in a given chromosome, directions of the DMRs (hyper- or hypo-), and the number of CpGs 
in the DMRs. 

## Functions

The hummingbird package contains the following three functions: 

(1) hummingbirdEM: Reads the input data, sets the initial values, executes the Expectation-Maximization 
(EM) algorithm, an estimation procedure for the Bayesian HMM and infers the best sequence of methylation 
states.

(2) hummingbirdPostAdjustment: Allows researchers to place extra requirements on the selection of DMRs, 
such as the minimum length of a DMR, the minimum number of CpGs in a DMR, and the maximum distance 
(in base pairs) between any two adjacent CpGs.

(3) hummingbirdGraph: Generates the Methylation Level and Prediction graphs for the user-specified regions.

## Sample Dataset

A sample dataset "exampleHummingbird" is provided with the package as an example. 

Specifically, partial data of chromosome 29 of the large offspring syndrome (LOS) study as described in "Chen, Z. et al (2017): Global misregulation of genes largely uncoupled to DNA methylome epimutations characterizes a congenital overgrowth syndrome. Scientific Reports 7, 12667". The raw FASTQ files of the WGBS experiment from this study are publibly available at Gene Expression Omnibus with accession no. GSE93775.

The matrices abnormUM, abnormM, normM, normUM and pos are given as a SummarizedExperiment input object to the functions of the hummingbird package. They are the methylation data from the WGBS experiment. In this experiment, there are four replicates in the case (LOS) group and four replicates in the control group. However, this is just an example. The hummingbird package does not require replication in either the control or the case group.

The outputs of the hummingbirdEM, hummingbirdPostAdjustment and the hummingbirdGraph functions when the aforementioned matrices are used are described in the Example section.

Below is a short presentation of the data:

```{r sampleDataset}
library(GenomicRanges)
library(SummarizedExperiment)
library(hummingbird)
data(exampleHummingbird)
```

The CpG positions:
```{r sampleDataset_pos}
rowRanges(exampleSE)
```

The matrices containing the methylated and unmethylated read count data of the normal group. Each column of the matrix represents a replicate and each row represents a CpG position:
```{r sampleDataset_norm}
head(assays(exampleSE)[["normM"]])
head(assays(exampleSE)[["normUM"]])
```

The matrices containing the methylated and unmethylated read count data of the abnormal group. Each column of the matrix represents a replicate and each row represents a CpG position:
```{r sampleDataset_abnorm}
head(assays(exampleSE)[["abnormM"]])
head(assays(exampleSE)[["abnormUM"]])
```


## Example

Loading the hummingbird package and the exampleHummingbird dataset:

```{r hummingbird}
library(hummingbird)
data(exampleHummingbird)
```

- hummingbirdEM: Expectation-Maximization Algorithm for Fitting the Hidden Markov Model

This function reads in methylated and unmethylated read count data, transforms it into logarithm bin-wise data, sets up initial values and implements the EM algorithm to estimate HMM parameters and find the best sequence of hidden states based on model fitting.

emInfo is the output, a GenomicRanges object containing the start and end positions of each bin, the distance between the current bin the bin ahead of it, the average methylation rate of abnormal and normal groups and the initial direction.

```{r hummingbird_em}
emInfo <- hummingbirdEM(experimentInfo = exampleSE, binSize = 40)
emInfo
```

- hummingbirdPostAdjustment: Post Adjustment algorithm for the output of the EM

As input to the function, the output emInfo of the hummingbirdEM, the pos matrix and user-defined parameters are given.

This function adjusts HMM output such that each detected DMR has a (user-defined) minimum length, maximum gap and minimum number of CpGs in each DMR.

postAdjInfo is the output, a list of two GenomicRanges objects, the obsPostAdj and the DMRs. The DMRs contains the detected regions based on the user-defined arguments.The start genomic position, the end genomic position, Length of region, Direction ("0" indicates no significant change, "1" indicates predicted hyper-methylation, and "2" indicates predicted hypo-methylation) and the Number of CpGs.

```{r hummingbird_postAdjustment}
postAdjInfo <- hummingbirdPostAdjustment(experimentInfo = exampleSE, emInfo = emInfo, minCpGs = 10, minLength = 100, maxGap = 300)
postAdjInfo$DMRs
```

- hummingbirdGraph: Generates the Methylation Level and Prediction graphs for the user-specified regions.

The Methylation Level graph is the sample averages from two comparison groups. 

The Predictions graph is the sample average difference along with predictions, where "0" indicates no significant change, "1" indicates predicted hyper-methylation, and "2" indicates predicted hypo-methylation.

In the next figures, in order to visualize the 2nd DMR, which contains the largest number of CpGs (12), its coordinates (as specified in postAdjInfo$DMRs) are entered by the user:

```{r hummingbird_graph}
hummingbirdGraph(experimentInfo = exampleSE, postAdjInfoDMRs = postAdjInfo$DMRs, coord1 = 107991, coord2 = 108350)
```

## Citation

If you use the hummingbird package, please cite the following paper:

- Ji T. A Bayesian hidden Markov model for detecting differentially methylated regions. Biometrics. 2019;75(2):663‐673. DOI:10.1111/biom.13000

The paper includes the detailed information of the algorithm.


